src/kgen.cpp

```diff
@@ -1,113 +1,128 @@
 #include <iostream>
 #include <vector>
 #include <algorithm>
+#include <functional>
+
+namespace kgen {
+
+    template<typename T>
+    class gen {
+    private:
+        class _lookback {
+        protected:
+            int ctr;
+        public:
+            explicit _lookback(int i) : ctr{i} {}
+
+            void bump() {
+                ++ctr;
+            }
+        };
+
+        std::vector<std::reference_wrapper<_lookback>> lbs;
+        bool init = false;
+        T val;
+
+    protected:
+        template<typename U, int Max>
+        class lookback : public _lookback {
+        public:
+            std::vector<U> circular_buf;
+
+            lookback() : _lookback(0), circular_buf(Max, U{}) {}
+
+            explicit lookback(const U &init) : _lookback(0), circular_buf(Max, init) {}
+
+            lookback(lookback &l) : _lookback(l.ctr), circular_buf(l.circular_buf) {}
+
+            lookback &operator=(const U &val) {
+                circular_buf[_lookback::ctr % Max] = val;
+                return *this;
+            }
+
+            U &operator[](int i) {
+                if (i > 0)
+                    throw std::invalid_argument("It's not called lookback not lookahead!");
+                if (-i > Max)
+                    throw std::invalid_argument(
+                            "Can't lookback more than " + std::to_string(Max) + " (attempted: " + std::to_string(-i) +
+                            ")");
+                return circular_buf[(_lookback::ctr + i + Max) % Max];
+            }
+
+            U &operator*() {
+                return circular_buf[_lookback::ctr % Max];
+            }
+        };
+
+        virtual T next() = 0;
+
+        gen(std::initializer_list<std::reference_wrapper<_lookback>> _lbs)
+                : lbs{_lbs} {}
+
+    public:
+        typedef std::input_iterator_tag iterator_category;
```

Comment:
what is this for?

Reply:
I figured it would be good practice to provide that semantic for any algorithm that requires an input iterator.